### PR TITLE
[netflow] Add Ephemeral Port Rollup

### PR DESCRIPTION
**Pull request by AlexandreYang**
### What does this PR do?

[netflow] Add Ephemeral Port Rollup

### Motivation

Pros:
- Less aggregated flows and less flows sent to Datadog means a Single Agent can support more flows per sec and more devices (which is an important goal since we want users to use a Single Agent as much as possible)
- Less memory usage due to better aggregation ratio, less flows are stored during aggregation
- Less flows stored in Event Store => Better query latency

Cons:
- Ephemeral port rollups can lead to port data being lost. The assumption we are making is that those Ports (ephemeral) have no/low value to customers.
- In case user find the default ephemeral port aggregation threshold too aggressive, they can change aggregator_port_rollup_threshold.
- Setting very high number to aggregator_port_rollup_threshold is equivalent to disabling ephemeral port rollup.

### Describe how to test/QA your changes

1/ Test we are roll up works on mimic: random ports are aggregated as expected

2/ Test Port Rollup `aggregator_port_rollup_threshold`

3/ Test `aggregator_rollup_tracker_refresh_interval`

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/netflow/config/config.go [hunk 1]
```diff
@@ -78,6 +81,12 @@ func ReadConfig() (*NetflowConfig, error) {
 	if mainConfig.AggregatorBufferSize == 0 {
 		mainConfig.AggregatorBufferSize = common.DefaultAggregatorBufferSize
 	}
+	if mainConfig.AggregatorPortRollupThreshold == 0 {
+		mainConfig.AggregatorPortRollupThreshold = common.DefaultAggregatorPortRollupThreshold
+	}
+	if mainConfig.AggregatorRollupTrackerRefreshInterval == 0 {
```
[the remark this answers is not in the document]
Comment (pducolin) [reply pkg/netflow/config/config.go:87]: What about using the flush interval ?

File: pkg/netflow/flowaggregator/aggregator.go [hunk 2]
```diff
@@ -137,3 +145,8 @@ func (agg *FlowAggregator) flush() int {
 
 	return len(flowsToFlush)
 }
+
+func (agg *FlowAggregator) rollupTrackersRefresh() {
+	log.Debugf("Rollup tracker refresh: use new store as current store")
+	agg.flowAcc.portRollup.UseNewStoreAsCurrentStore()
```
Comment (pducolin) [inline pkg/netflow/flowaggregator/aggregator.go:151]: **💬 suggestion**
I know I know, I am usually the one in love with long names, but what about calling this just `Refresh` or `Reset` ?

File: pkg/netflow/flowaggregator/flowaccumulator.go [hunk 3]
```diff
@@ -85,15 +94,25 @@ func (f *flowAccumulator) flush() []*common.Flow {
 }
 
 func (f *flowAccumulator) add(flowToAdd *common.Flow) {
-	f.mu.Lock()
-	defer f.mu.Unlock()
+	log.Tracef("Add new flow: %+v", flowToAdd)
+
+	// Handle port rollup
+	f.portRollup.Add(flowToAdd.SrcAddr, flowToAdd.DstAddr, uint16(flowToAdd.SrcPort), uint16(flowToAdd.DstPort))
+	portCount, isEphemeralSource := f.portRollup.GetPortCount(flowToAdd.SrcAddr, flowToAdd.DstAddr, uint16(flowToAdd.SrcPort), uint16(flowToAdd.DstPort))
```
Comment (pducolin) [inline pkg/netflow/flowaggregator/flowaccumulator.go:101]: **💬 suggestion**
Rather then returning a port count and a boolean, I would return a sourcePortCount and a destinationPortCount. Or, even better, return two booleans, or an enum, telling if this is an ephemeralSourcePort, an ephemeralDestPort, or a non ephemeral conversation. To me `portRollup` should handle the logic of the threshold, sounds like you are mixing business boundaries here

File: pkg/netflow/portrollup/portrollup.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// NewEndpointPairPortRollupStore create a new *EndpointPairPortRollupStore
+func NewEndpointPairPortRollupStore(portRollupThreshold int) *EndpointPairPortRollupStore {
+	return &EndpointPairPortRollupStore{
+		// curStore and newStore map key is composed of `<SOURCE_IP>|<DESTINATION_IP>`
+		// SOURCE_IP and SOURCE_IP are converted from []byte to string.
+		// string is used as map key since we can't use []byte as map key.
+		curStore: make(map[string]*portRollupTracker),
+		newStore: make(map[string]*portRollupTracker),
```
Comment (pducolin) [inline pkg/netflow/portrollup/portrollup.go:30]: **💭 thought**
Can you add a comment about why we need a second store ? Took me some time to understand you need it because you reset it every `interval`, and you need to keep the past `interval` value to avoid resetting to 0. Also, this means that you are actually rolling up over a 2 * interval period, right ?

File: pkg/netflow/portrollup/portrolluptracker.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	addPort(pr.sourcePorts, sourcePort, destPort, portRollupThreshold)
+	addPort(pr.destPorts, destPort, sourcePort, portRollupThreshold)
+}
+
+func addPort(ports map[uint16]map[uint16]bool, port uint16, ephemeralPort uint16, portRollupThreshold int) {
+	if _, ok := ports[port]; !ok {
+		ports[port] = make(map[uint16]bool)
+	}
+
+	if len(ports[port]) < portRollupThreshold && !ports[port][ephemeralPort] {
```
Comment (pducolin) [inline pkg/netflow/portrollup/portrolluptracker.go:51]: **❓ question**
Shouldn't this be `len(ports[port]) > portRollupThreshold` ?

**Review by pducolin — CHANGES_REQUESTED** [review]
I would review the interface to the rollup actor. I like how you handle the threshold from there, I would expect it to return only the information you need, so if one or the other port is ephemeral. No need to return the count.

File: pkg/netflow/flowaggregator/aggregator.go (hunk 2 above)
Comment (AlexandreYang) [reply pkg/netflow/flowaggregator/aggregator.go:151]: For this case, I think `Refresh` or `Reset` might not be suitable since since usually idempotent (can be call multiple time with same result), but in this `UseNewStoreAsCurrentStore` have a very precise operation that can't be called multiple times, and I would favour being explicit here.

File: pkg/netflow/portrollup/portrolluptracker.go (hunk 5 above)
Comment (AlexandreYang) [reply pkg/netflow/portrollup/portrolluptracker.go:51]: It will work too, but I would like to avoid writing `ports[port][ephemeralPort] = true` when the value is already `true`, hence the use of `&& !ports[port][ephemeralPort]`.

File: pkg/netflow/common/utils.go [hunk 6]
```diff
@@ -23,6 +23,14 @@ func MaxUint64(a uint64, b uint64) uint64 {
 	return b
 }
 
+// MaxUint16 returns the max of the two passed number
+func MaxUint16(a uint16, b uint16) uint16 {
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply pkg/netflow/common/utils.go:27]: > Why not using math.Max ? Because of conversion to float ?

Yes, that's right. 

> This sounds like a function other teams could use, do we have any common math utils ? Or math package extender ?

I'm not seeing any common math utils in `datadog-agent`. There are external libraries, but I'm not sure it's worth importing them simple function like this one.

With Go, 1.18, we can possibly use generics instead: https://stackoverflow.com/a/27516559

File: pkg/netflow/config/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply pkg/netflow/config/config.go:87]: > Challenging this: what is the use case for this option ? Isn't AggregatorPortRollupThreshold to fine tune this ? I wonder if it adds not necessary complexity.

This TTL option is use to define how often we should `UseNewStoreAsCurrentStore`, the interval we should drop old tracked ports (we actually keep port data `2 * AggregatorRollupTrackerRefreshInterval` seconds ).

> What about using the flush interval ?

If AggregatorRollupTrackerRefreshInterval is too short, ephemeral port rollup might be flaky if the port count is around the threshold.

I think those are two different intervals that should have different TTL.

File: pkg/netflow/portrollup/portrolluptracker.go (hunk 5 above)
Comment (pducolin) [reply pkg/netflow/portrollup/portrolluptracker.go:51]: Let me rephrase, shouldn't this be 
```suggestion
	if len(ports[port]) > portRollupThreshold && !ports[port][ephemeralPort] {
```

File: pkg/netflow/flowaggregator/aggregator.go (hunk 2 above)
Comment (pducolin) [reply pkg/netflow/flowaggregator/aggregator.go:151]: Wonder if the timer should be inside `portRollup`. You pass the observation window, and it will take care of dropping old data.

File: pkg/netflow/portrollup/portrolluptracker.go (hunk 5 above)
Comment (AlexandreYang) [reply pkg/netflow/portrollup/portrolluptracker.go]: I see, thanks for rephrasing. It's actually related to:

> 	// we only track additional ports if portRollupThreshold has not been reached
>	// if portRollupThreshold is already reached, there is no value to track more ports
>	// since we already know we need to rollup

But since the check is already made here: https://github.com/DataDog/datadog-agent/pull/13367/files/8000e6a18acce8b6ecd55f6da5509ac09828d348#diff-544bcf72f2eff2bc13439925d88e223b59ba3df2e265388bbeddee5041eb4d9aR38-R40

There is no need to do an additional check here: Removed here https://github.com/DataDog/datadog-agent/pull/13367/commits/a044fea914964b59f13a07b1246b5015ccdfd253

File: pkg/netflow/flowaggregator/aggregator.go (hunk 2 above)
Comment (AlexandreYang) [reply pkg/netflow/flowaggregator/aggregator.go:151]: I don't really agree with the last point :), I think it's good to keep the timer/scheduling part outside to keep `portRollup`. Usually, adding timing to portRollup makes it less reusable, reliant on time, and harder to test.

File: pkg/netflow/flowaggregator/flowaccumulator.go (hunk 3 above)
Comment (AlexandreYang) [reply pkg/netflow/flowaggregator/flowaccumulator.go]: Fixed here https://github.com/DataDog/datadog-agent/pull/13367/commits/168af5b98ab357b7fbce83fc959a9a41280df616

**Review by gmarabout — APPROVED** [review]
I love it!

File: pkg/netflow/portrollup/portrollup.go (hunk 4 above)
Comment (AlexandreYang) [reply pkg/netflow/portrollup/portrollup.go:45]: Good idea. Added here 144607ee3

**Review by pducolin — APPROVED** [review]
Thank you for addressing comments. Well done !

**Comment by AlexandreYang** [comment]
@DataDog/documentation can I have a review for this PR ? :)

**Review by kayayarai — APPROVED** [review]
